File: backend/src/orders/core/Order.ts
import type { OrderItem } from "./OrderItem.ts";
import type { Payment } from "./Payment.ts";

/**
 * Represents an Order domain entity.
 */
export class Order {
  /**
   *
   */
  constructor(
    public readonly id: string,
    public readonly customerId: string | null,
    public readonly guestEmail: string | null,
    public readonly guestName: string | null,
    public readonly guestPhone: string | null,
    public readonly shippingAddress: string | null, // JSON string
    public readonly subtotal: number, // in cents
    public readonly total: number, // in cents
    public readonly currency: string,
    public readonly status:
      "pendiente" | "pagado" | "enviado" | "entregado" | "cancelado" | "fallido",
    public readonly stripePaymentIntentId: string | null,
    public readonly notes: string | null,
    public readonly createdAt: string,
    public readonly updatedAt: string,
    public readonly items: OrderItem[] = [],
    public readonly payments: Payment[] = [],
  ) {}

  /**
   *
   */
  public static create(params: {
    id: string;
    customerId?: string | null;
    guestEmail?: string | null;
    guestName?: string | null;
    guestPhone?: string | null;
    shippingAddress?: string | null;
    subtotal: number;
    total: number;
    currency?: string;
    status?: "pendiente" | "pagado" | "enviado" | "entregado" | "cancelado" | "fallido";
    stripePaymentIntentId?: string | null;
    notes?: string | null;
    createdAt?: string;
    updatedAt?: string;
    items?: OrderItem[];
    payments?: Payment[];
  }): Order {
    const now = new Date().toISOString();
    return new Order(
      params.id,
      params.customerId ?? null,
      params.guestEmail ?? null,
      params.guestName ?? null,
      params.guestPhone ?? null,
      params.shippingAddress ?? null,
      params.subtotal,
      params.total,
      params.currency ?? "PEN",
      params.status ?? "pendiente",
      params.stripePaymentIntentId ?? null,
      params.notes ?? null,
      params.createdAt ?? now,
      params.updatedAt ?? now,
      params.items ?? [],
      params.payments ?? [],
    );
  }
}

File: mobile/lib/features/orders/domain/entities/order.dart
import 'package:mobile/features/orders/domain/entities/order_item.dart';
import 'package:mobile/features/orders/domain/entities/payment.dart';

/// Order domain entity on mobile.
class Order {
  final String id;
  final String? customerId;
  final String? guestEmail;
  final String? guestName;
  final String? guestPhone;
  final String? shippingAddress; // JSON string
  final int subtotal; // in cents
  final int total; // in cents
  final String currency;
  final String
  status; // 'pendiente', 'pagado', 'enviado', 'entregado', 'cancelado', 'fallido'
  final String? stripePaymentIntentId;
  final String? notes;
  final String createdAt;
  final String updatedAt;
  final List<OrderItem> items;
  final List<Payment> payments;

  const Order({
    required this.id,
    this.customerId,
    this.guestEmail,
    this.guestName,
    this.guestPhone,
    this.shippingAddress,
    required this.subtotal,
    required this.total,
    required this.currency,
    required this.status,
    this.stripePaymentIntentId,
    this.notes,
    required this.createdAt,
    required this.updatedAt,
    required this.items,
    required this.payments,
  });
}
